receipt_listener.go

```diff
@@ -74,149 +82,154 @@ func (l *ReceiptListener) Run(ctx context.Context) error {
+				r := receipt(common.BytesToHash(log.Data))
+				r.Results = append(r.Results, &MetaTxnResult{
+					Status: MetaTxnExecuted,
+				})
+			} else if len(log.Topics) == 1 && log.Topics[0] == TxFailedEventSig {
+				// definite TxFailed event
+
+				metaTxnID, reason, err := DecodeTxFailedEvent(log)
+				if err != nil {
+					l.log.Err(err).Msgf("unable to decode TxFailed event: topics=%v data=%v", log.Topics, log.Data)
+					continue
+				}
+
+				r := receipt(metaTxnID)
+				r.Results = append(r.Results, &MetaTxnResult{
+					Status: MetaTxnFailed,
+					Reason: reason,
+				})
 			}
+		}
 
-			// Add found result to block of receipts
-			blockOfReceipts = append(blockOfReceipts, result)
+		receipts[txHash] = make([]*ReceiptResult, 0, len(metaTxnReceipts))
+		for _, receipt := range metaTxnReceipts {
+			receipts[txHash] = append(receipts[txHash], receipt)
 		}
 	}
 
-	// Nothing to record, skipping.
-	if len(blockOfReceipts) == 0 {
-		return nil
+	for txHash, txReceipts := range receipts {
+		l.handleReceipts(txHash, txReceipts)
+	}
+}
+
+func (l *ReceiptListener) handleReceipts(txHash common.Hash, txReceipts []*ReceiptResult) {
+	if len(txReceipts) == 0 {
+		return
 	}
 
-	// Publish to subscribers
-	l.mu.Lock()
-	defer l.mu.Unlock()
-
-	for _, result := range blockOfReceipts {
-		for _, sub := range l.subscribers {
-			select {
-			case <-sub.done:
-			case sub.ch <- result:
-			case <-time.After(2 * time.Second):
-				l.log.Warn().Msgf("channel publisher is blocked by a slow subscriber")
+	// handle receipts in an independent goroutine so that node failures don't stall the block handler
+	go func() {
+		err := l.br.Do(context.Background(), func() error {
+			receipt, err := l.provider.TransactionReceipt(context.Background(), txHash)
+			if err != nil {
+				return fmt.Errorf("unable to fetch receipt for %v: %w", txHash.Hex(), err)
+			} else if receipt == nil {
+				return fmt.Errorf("unable to fetch receipt for %v", txHash.Hex())
 			}
+
+			// populate TxnReceipt field
+			for _, txReceipt := range txReceipts {
+				txReceipt.TxnReceipt = receipt
+			}
+
+			return nil
+		})
+		if err != nil {
+			l.log.Warn().Err(err).Msgf("failed to fetch receipt after several tries")
 		}
-	}
 
-	l.log.Debug().
-		Int("past-block-entries", len(l.pastReceipts)).
-		Int("new-entries", len(blockOfReceipts)).
-		Msgf("Push into past receipts")
+		l.muSubscribers.Lock()
+		defer l.muSubscribers.Unlock()
 
-	if len(l.pastReceipts) < 1024 {
-		// Append at the end of slice
-		l.pastReceipts = append(l.pastReceipts, blockOfReceipts)
-	} else {
-		// Append value but also pop the queue
-		l.pastReceipts = append(l.pastReceipts[1:], blockOfReceipts)
-	}
+		// broadcast receipts
+		for _, txReceipt := range txReceipts {
+			for _, sub := range l.subscribers {
+				select {
+				case <-sub.done:
+				case sub.sendCh <- txReceipt:
+				}
+			}
+		}
 
-	return nil
+		if len(l.pastReceipts) < 4096 {
+			l.pastReceipts = append(l.pastReceipts, txReceipts)
```

Comment:
also .. we should probably make a separate method on ReceiptListener called `pushReceipt` and here we should have a lock too, since we have multiple go-routines writing to the same slice, we'll need a lock in that method

Reply:
you know, I think we are okay here because we have lock above it a few lines above on the subscriber lock..